kong/db/migrations/core/004_110_to_120.lua

```diff
@@ -0,0 +1,39 @@
+return {
+  postgres = {
+    up = [[
+      DO $$
+      BEGIN
+        ALTER TABLE IF EXISTS ONLY "routes" ADD "https_redirect_status_code" INTEGER DEFAULT 426;
```

Comment:
It is known that altering a table by adding a new column with a default value will lock the table (if it is large enough, maybe for a significant amount of time). There are many resources describing the issue online, you probably have heard of it yourself before. http://www.joshuakehn.com/2017/9/9/postgresql-alter-table-and-long-transactions.html

Considering the potential size of the routes table (up to a few thousand rows), combined with our B/G migrations path, it seems we should give some extra attention to a possible scenario, in which new green nodes are starting up and attempting to query the routes table (to build their router) at the same time the physical rewrite is still taking place in Postgres.
Another argument can be made in favour of writing referential code in our core migrations, since we have learned a long time ago that core patterns tend to spread in tangential code (e.g. plugins). It is not unrealistic to expect future similar migrations (adding new columns) to other tables, potentially much larger ones (e.g. oauth2 access tokens). Writing best-practices migrations for core could help preventing issue down the road.

Reply:
@hbagdi If you set declare a field in the schema with a `default` value AND `required = true` (to forbid `null` as a valid value), the schema library will fill in the default for you with no need to set it in the migration.

Reply:
> If you set declare a field in the schema with a default value AND required = true (to forbid null as a valid value), the schema library will fill in the default for you with no need to set it in the migration.

Is this something we have done in the past and is it safe to do so?
Imagine we take the approach, now, the value will not be set in the database until there is an update operation for routes. And if the default value changes (unlikely in this case, but for sake of broader discussion), then the value of field will change across upgrades, assuming the user has not updated the route between migrations, right?

Reply:
> Considering the potential size of the routes table (up to a few thousand rows), combined with our B/G migrations path, it seems we should give some extra attention to a possible scenario, in which new green nodes are starting up and attempting to query the routes table (to build their router) at the same time the physical rewrite is still taking place in Postgres.

I did consider this problem but considering there would be a few thousand routes or tens of thousand entries, this operation would be fairly quick on any modern hardware. I do understand the concern around how this pattern can (and will) spread and we should do something about it.

@thibaultcha, what do you propose in that case? I'm thinking of writing a migration similar to the Cassandra migration where we update the rows individually.

Reply:
@hbagdi My thinking was to set the default on Lua-land (e.g. in the proxy code itself `ngx.status = route.https_redirect_status_code or 426`. I _seem_ to recall that we followed the schema approach suggested by @hishamhm before, but I can't remember where. It's a good approach too imho, in this case I don't see the default value changing anytime soon, and we can worry about the default value changing if the issue ever arises. Most likely though, the "old default" value will still be accepted (backwards-compatibility), and if the "new default" changed, it's probably fine as per the new code. Now up to us to instruct users on how to best handle blue nodes, which we'll assess when and if the time comes.

Reply:
> My thinking was to set the default on Lua-land (e.g. in the proxy code itself ngx.status = route.https_redirect_status_code or 426

Right, so I already added this piece to handle the C* case because v1.2 nodes would come up and start before we finish writing the default value in the column and the new column would be nil on v1.2 nodes. This was before @hishamhm's comment on using `default` along with `required`.

Here is what we could do:
1. We add the column in both Postgres (no default value) and Cassandra during migration.
2. We do not write the default value to the DB during the `teardown` process (I'm not sure if I'm implying this correctly from the above comment or not, so please let me know if incorrect).
3. We add `required = true` in the schema so that the it is always filled in with the default value.

Reply:
@hbagdi :+1: yes, that's the approach we've been taking in Core with this kind of migration (I also don't remember where exactly we did this, it was most likely in some plugin — but it was definitely something that was part of a conversation on B/G-friendly migration patterns). It avoids having to hardcode the default in proxy code and keeps the schema as source of truth. (Plus, note that letting the schema fill the default is no better or worse than doing it in Lua-land as far as "what if the defaults change" is concerned.)